internal/website/content/howto/pubsub/subscribe.md

```diff
@@ -0,0 +1,504 @@
+[Azure Service Bus library][] and then pass the subscription to
+`azuresb.OpenSubscription`. There are also helper functions in the `azuresb`
+package to make this easier.
+
+```go
+import (
+    "os"
+
+    "gocloud.dev/pubsub/azuresb"
+)
+
+// ...
+
+// Change these as needed for your application.
+serviceBusConnString := os.Getenv("SERVICEBUS_CONNECTION_STRING")
+const topicName = "test-topic"
+const subscriptionName = "test-subscription"
+
+// Connect to Azure Service Bus for the given subscription.
+busNamespace, err := azuresb.NewNamespaceFromConnectionString(serviceBusConnString)
+if err != nil {
+    return err
+}
+busTopic, err := azuresb.NewTopic(busNamespace, topicName, nil)
+if err != nil {
+    return err
+}
+defer busTopic.Close(ctx)
+busSub, err := azuresb.NewSubscription(busTopic, subscriptionName, nil)
+if err != nil {
+    return err
+}
+defer busSub.Close(ctx)
+
+// Construct a *pubsub.Subscription.
+subscription, err := azuresb.OpenSubscription(ctx, busSub, nil)
+if err != nil {
+    return err
+}
+defer subscription.Shutdown(ctx)
+```
+
+[`azuresb.OpenSubscription`]: https://godoc.org/gocloud.dev/pubsub/azuresb#OpenSubscription
+[Azure Service Bus library]: https://github.com/Azure/azure-service-bus-go
+
+## RabbitMQ {#rabbitmq}
+
+The Go CDK can receive messages from an [AMQP 0.9.1][] queue, the dialect of
+AMQP spoken by [RabbitMQ][]. A RabbitMQ URL only includes the queue name.
+The RabbitMQ's server is discovered from the `RABBIT_SERVER_URL` environment
+variable (which is something like `amqp://guest:guest@localhost:5672/`).
+
+```go
+import (
+    "gocloud.dev/pubsub"
+    _ "gocloud.dev/pubsub/rabbitpubsub"
+)
+
+// ...
+
+subscription, err := pubsub.OpenSubscription(ctx, "rabbit://myqueue")
+if err != nil {
+    return err
+}
+defer subscription.Shutdown(ctx)
+```
+
+[AMQP 0.9.1]: https://www.rabbitmq.com/protocol.html
+[RabbitMQ]: https://www.rabbitmq.com
+
+### RabbitMQ Constructor {#rabbitmq-ctor}
+
+The [`rabbitpubsub.OpenSubscription`][] constructor opens a RabbitMQ queue.
+You must first create an [`*amqp.Connection`][] to your RabbitMQ instance.
+
+```go
+import (
+    "github.com/streadway/amqp"
+    "gocloud.dev/pubsub/rabbitpubsub"
+)
+
+// ...
+
+rabbitConn, err := amqp.Dial("amqp://guest:guest@localhost:5672/")
+if err != nil {
+    return err
+}
+subscription := rabbitpubsub.OpenSubscription(rabbitConn, "myqueue", nil)
+defer subscription.Shutdown(ctx)
+```
+
+[`*amqp.Connection`]: https://godoc.org/github.com/streadway/amqp#Connection
+[`rabbitpubsub.OpenSubscription`]: https://godoc.org/gocloud.dev/pubsub/rabbitpubsub#OpenSubscription
+
+## NATS {#nats}
+
+The Go CDK can publish to a [NATS][] subject. A NATS URL only includes the
+subject name. The NATS server is discovered from the `NATS_SERVER_URL`
+environment variable (which is something like `nats://nats.example.com`).
+
+**TODO(light): Something about the `ackfunc` query parameter? Unsure when to use.**
```

Comment:
NATS is an "at most once" system. The moment you (the subscriber) receive a message, it will never be delivered again. So it's meaningless to ack it.

We talked a lot about what the behavior of `Message.Ack` should be in this case. I argued that if your code calls `Ack`, you think you are running an at-least-once system, and you expect it to be fault tolerant. (If the subscriber crashes before acking, the message will be retransmitted, so every message will eventually be processed.) Therefore, goes this argument, `Ack` should panic, because you have committed the grievous error of linking in at at-most-once provider with your at-least-once system.

@vangent  argued that panicking is going to make some forms of testing very tedious. You may want to run in production with an at-least-once system, but test and do local development with a cheap, on-box system like NATS. To this way of thinking, `Ack` should be a no-op.

Since neither of us could convince the other, we punted to the user. They can (and must) tell us what `Ack` means to them, by providing a function to call when `Message.Ack` is invoked. It is "must," not "can," (i.e. there is no default), as a concession to me. I don't want anyone accidentally getting it wrong (cf. "grievous error" above).

We thought we explained this in [the package doc](https://godoc.org/gocloud.dev/pubsub#hdr-At_most_once_and_At_least_once_Delivery). Clearly we failed. Please suggest alternative wording.

Reply:
> NATS is an "at most once" system. The moment you (the subscriber) receive a message, it will never be delivered again. So it's meaningless to ack it.
> 
> We talked a lot about what the behavior of `Message.Ack` should be in this case. I argued that if your code calls `Ack`, you think you are running an at-least-once system, and you expect it to be fault tolerant. (If the subscriber crashes before acking, the message will be retransmitted, so every message will eventually be processed.) Therefore, goes this argument, `Ack` should panic, because you have committed the grievous error of linking in an at-most-once provider with your at-least-once system.
> 
> @vangent argued that panicking is going to make some forms of testing very tedious. You may want to run in production with an at-least-once system, but test and do local development with a cheap, on-box system like NATS. To this way of thinking, `Ack` should be a no-op.

With you so far.

> Since neither of us could convince the other, we punted to the user.

Bluntly, I don't think a user could ever make this decision, at least not with the information provided. (See below.)

> They can (and must) tell us what `Ack` means to them, by providing a function to call when `Message.Ack` is invoked. It is "must," not "can," (i.e. there is no default), as a concession to me. I don't want anyone accidentally getting it wrong (cf. "grievous error" above).

Minor note: the reference doesn't make this clear. I assumed that one of these query parameter values is the default and that the default isn't documented.

> We thought we explained this in [the package doc](https://godoc.org/gocloud.dev/pubsub#hdr-At_most_once_and_At_least_once_Delivery). Clearly we failed. Please suggest alternative wording.

I don't think that I can make such a suggestion because I don't know the information that would help me (or a user) make such a decision.

Concretely, what I am missing is the "why?" not the "what?" or "how?" of each of these options. I understand mechanically what each of these ack function options will do, but I do not understand the tradeoffs associated with picking any one of the individual choices. If I were to come across this cold, I probably would be lazy and pick no-op because that sounds the least dangerous. However, if I were given more information as to "picking no-op is useful in scenarios where X but is dangerous if Y: bad thing Z will happen", that would be tremendously helpful.

Reply:
> I assumed that one of these query parameter values is the default and that the default isn't documented.

This is a query parameter without a default, and indeed that should be clearly noted. I love URLs!\</sarcasm>

Reply:
> I do not understand the tradeoffs associated with picking any one of the individual choices.

If you pick no-op, you could be breaking an invariant of your system, causing the system to malfunction (but probably only rarely). Here's the scenario:

1. You design your system to include an at-least-once pubsub provider. You chose at-least-once for reliability: you *must* process every message, and you know that if one of your subscribers crashes or locks up before acking, the message will be redelivered to someone else. You use the `gcloud.dev/pubsub` package and link in one of the at-least-once providers.

2. One year later, a maintainer of your system notices that the company's cloud provider just introduced another pubsub product. The maintainer doesn't read the fine print, which says that this is an at-most-once system, but does notice two numbers: it's a lot faster than the pubsub provider you're using, and it's a lot cheaper too.

3. Seems like a no-brainer! The maintainer swaps out the pubsub provider you chose in step 1 for the cheaper one, which also has a Go CDK driver. When they code up the `OpenSubscription` call of the new provider, they see they  must provide an ack function. Let's say they choose either no-op or panic.

4.
    a. They choose no-op. They build and redeploy. Everything seems fine (but faster and cheaper!). But a week later, they start seeing customer support issues: "I heard that XYZ happened but your app didn't notify me." Not a lot, maybe just one or two out of millions of customers. These support calls trickle in for months. They happen so rarely that they are hard to follow up on, and everyone is busy with adding new features. Eventually, someone notices the correlation between delivery failure and crashed machines. Some time later, someone figures out that if a machine crashes before it handles a pubsub message, the message is never re-delivered. Eventually they track it down to the change in step 3.

   b. They choose panic. They build and redeploy. Almost immediately, their monitoring system goes through the roof: processes are panicking everywhere. Multiple engineers are forced to look at the code causing the panic. Most of them suggest swapping out the panic for a no-op, but someone suggests thinking about what's going on a bit more. Disaster averted.

On the other hand, if you pick no-op, you can test your system locally with a cheaper and simpler at-most-once system, without having to have different code (Ack vs. no Ack) running in testing and production.

(Caveat: There may be some bias in my presentation of the alternatives.)

Reply:
Thanks for the detailed write-up; I understand much better now. I went with recommending `?ackfunc=panic` as the first choice, and detailed under what conditions someone may choose `?ackfunc=noop`, which seems consistent with what you are saying.